packages/rum-recorder/src/domain/segmentCollection.spec.ts

```diff
@@ -0,0 +1,211 @@
+import { makeMouseMoveRecord } from '../../test/utils'
+import { IncrementalSource, MouseMoveRecord, Record, RecordType, SegmentContext, SegmentMeta } from '../types'
+import {
+  MAX_MOUSE_MOVE_BATCH,
+  MAX_SEGMENT_DURATION,
+  RecordsIncrementalState,
+  Segment,
+  SegmentWriter,
+  startSegmentCollection,
+} from './segmentCollection'
+
+class StringWriter implements SegmentWriter {
+  output = ''
+  completed: Array<{ meta: SegmentMeta; segment: SegmentMeta & { records: Record[] } }> = []
+  write(data: string) {
+    this.output += data
+  }
+  complete(data: string, meta: SegmentMeta) {
+    this.completed.push({ meta, segment: JSON.parse(this.output + data) as any })
+    this.output = ''
+  }
+}
+
+const CONTEXT: SegmentContext = { application: { id: 'a' }, view: { id: 'b' }, session: { id: 'c' } }
+const RECORD: Record = { type: RecordType.Load, timestamp: 10, data: {} }
+
+describe('startSegmentCollection', () => {
+  let writer: StringWriter
+
+  beforeEach(() => {
+    writer = new StringWriter()
+  })
+
+  afterEach(() => {
+    jasmine.clock().uninstall()
+  })
+
+  it('immediatly starts a new segment', () => {
+    const { addRecord } = startSegmentCollection(() => CONTEXT, writer)
+    expect(writer.output).toBe('')
+    addRecord(RECORD)
+    expect(writer.output).toBe('{"records":[{"type":1,"timestamp":10,"data":{}}')
+    expect(writer.completed.length).toBe(0)
+  })
+
+  it('writes a segment when renewing it', () => {
+    const { renewSegment, addRecord } = startSegmentCollection(() => CONTEXT, writer)
+    addRecord(RECORD)
+    renewSegment('before_unload')
+    expect(writer.completed.length).toBe(1)
+  })
+
+  it('writes a segment after MAX_SEGMENT_DURATION', () => {
+    jasmine.clock().install()
+    const { addRecord } = startSegmentCollection(() => CONTEXT, writer)
+    addRecord(RECORD)
+    jasmine.clock().tick(MAX_SEGMENT_DURATION)
+    expect(writer.completed.length).toBe(1)
+  })
+
+  it('does not write a segment after MAX_SEGMENT_DURATION if a segment has been created in the meantime', () => {
+    jasmine.clock().install()
+    const { renewSegment, addRecord } = startSegmentCollection(() => CONTEXT, writer)
+    addRecord(RECORD)
+    renewSegment('before_unload')
+    expect(writer.completed.length).toBe(1)
+    jasmine.clock().tick(MAX_SEGMENT_DURATION)
+    expect(writer.completed.length).toBe(1)
+  })
+
+  it("ignores calls to addRecord if context can't be get", () => {
```

Comment:
It can happen when `parentContext.findView` returns `undefined` of if the session id is `undefined` https://github.com/DataDog/browser-sdk/blob/benoit/implement-recorder/packages/rum-recorder/src/boot/recorder.ts#L51

Reply:
oh ok. I am wondering:

- should not we have a similar check that the one we are doing in assembly? (`session.isTracked() && viewContext && viewContext.session.id`)
- is there a reason to attach the context at the begin of the segment rather than on completion? it would be more similar to what we do on rum

Reply:
* `session.isTracked()` is indeed missing, I'll add it

* the segment context should not change during the segment lifetime, so getting it at the beginning or end should not change anything functionally. It would have drawbacks though:

  * When the view is created or session is renewed, `parentContext.findView()` and `session.isTracked()` are already set to the new values, so we would have an incorrect segment context in those cases. We could use `findView(segmentStartDate)`, but I don't think it is necessary. 

  * If we start a segment and fail to get the context at its end, the segment would be useless and we would have stored its records for nothing.